Opt-in method for purging (some) caches upon a QuotaExceededError

jeffposnick opened the pull request:
R: @philipwalton
CC: @raejin @Enalmada @josephliccini

Fixes #1308

There are two changes in this PR:

- There's a check in `workbox-core`'s `cacheWrapper.put()` (which is used "under the hood" throughout Workbox) method to detect when the underlying `cache.put()` fails due to a `QuotaExceededError`. When that happens, one or more previously registered callback functions will automatically be called, and given a chance to free up storage. There is (currently) no attempt made to retry the failed `cache.put()` automatically after the cleanup completes, and the `QuotaExceededError` will be re-thrown and bubble-up to the method that called `cacheWrapper.put()` originally.

- There's a new `purgeOnQuotaError` option added to the `workbox.expiration.Plugin` constructor. It defaults to `false`. When `true`, it will automatically register a quota error callback that will delete the **entire** cache (and IndexedDB) data that the plugin is responsible for. (This is based on #1500.)

Nothing should change by default, but developers who want to opt-in to trying this automatic deletion of caches should make sure that they're using `workbox.expiration.Plugin` for a given strategy, and set `purgeOnQuotaError: true` alongside their `maxEntries` and/or `maxAgeSeconds` configuration when constructing the plugin.

At some point in the future, we might:

- Automatically retry the failed `cache.put()` operation after all the callbacks are complete.
- Default to `purgeOnQuotaError: true` inside of `workbox.expiration.Plugin`.
- Add in some special logic inside `workbox.precaching` to handle quota errors in a different way.

In the meantime, I'd love to start getting some feedback from developers who frequently encounter quota issues with their real-world applications, and figure out how effective opting-in to this new behavior is.

It will have the most significant effect on developers whose runtime caches contribute the largest amounts towards their quota usage (perhaps because they're [caching opaque responses](https://stackoverflow.com/questions/39109789/what-limitations-apply-to-opaque-responses)). If you're precaching a massive amount of data, but not runtime caching much, this PR is not likely to help.

raejin commented:
Thank you for working on this! This is very exciting, as [the cache quota reporting chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=795701) was recently fixed. 

With this implementation, one is still possible to manually add logic to retry precaching assets when quota error callbacks are called.

philipwalton commented on packages/workbox-cache-expiration/Plugin.mjs:
```diff
@@ -235,7 +242,7 @@ class Plugin {
    */
   async deleteCacheAndMetadata() {
-    // Do this one at at a time instance of all at once via `Promise.all()` to
+    // Do this one at at a time instead of all at once via `Promise.all()` to
```
While we're fixing grammar, remove the duplicate "at" earlier in this line.

jeffposnick replied:
Done.

philipwalton commented on test/workbox-core/node/_private/test-quota.mjs:
```diff
@@ -0,0 +1,58 @@
+    sandbox.restore();
+  });
+
+  devOnly.it(`registerCallback() should throw when passed a non-function`, async function() {
```
Can you wrap all the tests related to `registerCallback()` in a describe block and then do a separate describe block for all the tests in `executeCallback()`.

jeffposnick replied:
I don't know about that—there's only one test that's just `registerCallback()` (which only runs in `devOnly` mode), and all of the rest of them rely on a combination of the two methods. That seems like overkill?

jeffposnick replied:
Done.

philipwalton commented on packages/workbox-core/_private.mjs:
```diff
@@ -15,23 +15,28 @@
-  DBWrapper,
   getFriendlyURL,
+  logger,
+  registerQuotaErrorCallback,
```
What's the reasoning behind exposing this?

jeffposnick replied:
Just flexibility.

While the automatic cleanup of caches that use expiration is one way of dealing with quota errors, you could imagine a web app that, for example, stores a large amount of data in IndexedDB. Developers who understand how their web app stores data could use this to register a callback that cleaned up that hypothetical IndexedDB database.

philipwalton commented on packages/workbox-cache-expiration/Plugin.mjs:
```diff
@@ -76,6 +79,10 @@ class Plugin {
     this._cacheExpirations = new Map();
+
+    if (config.purgeOnQuotaError) {
+      registerCallback(() => this.deleteCacheAndMetadata());
```
Is the plan that `registerCallback()` might/will get invoked in more places in the future? From what I can tell in this PR, this is the only place it's used, which makes the `quota.mjs` module seem like over-kill for just this.

jeffposnick replied:
Yes, along the lines of https://github.com/GoogleChrome/workbox/pull/1505#discussion_r194867665

In addition to giving developers a general-purpose way of registering a "cleanup" callback, we may want to make use of it elsewhere in the Workbox codebase—potentially as part of the precaching module, along the lines of #1312.

philipwalton replied:
Yeah, I get the intention of the change, I'm just wondering if it could be somehow better integrated into the rest of workbox.

Did you consider creating a new lifecycle callback: maybe something like `cacheWriteDidFail`? And in the same way that we can register a generic fetch failure handler, we could register generic cache failure handler.

Anyway, just a thought, I don't want to block this change if a better solution isn't obvious.

jeffposnick replied:
I felt like using the lifecycle callback naming conventions would be misleading, since here you're registering a callback that has a "global" (or `workbox.core`) scope, which will be invoked regardless of which strategy triggers the quota exception. That doesn't match the behavior of the lifecycle callbacks, which are all scoped to a given strategy.

`workbox.routing.setCatchHandler()` is closer in semantics, but that only supports setting a single callback, while we need support for multiple quota exception callbacks. To me, `register...` sounds more in line with being able to create multiple callbacks than `set...`, but I could change the naming if you feel strongly.

philipwalton replied:
> I felt like using the lifecycle callback naming conventions would be misleading, since here you're registering a callback that has a "global"...

Right, I wasn't suggesting just using the name, I was suggesting we actually add a `cacheWriteDidFail` lifecycle callback. While I understand that it's not global and would have to be added per strategy, that seems like a benefit to me, not a shortcoming. And users could write their strategy code in such a way that sharing the same callbacks across all their strategies isn't so bad.

My concern is that global handlers can lead to unpredictable behavior that's harder to test, especially for folks who are using Workbox in a more modular way, or anyone building a third-party plugin.

Anyway, I don't want to block this feature being added, especially if we know folks are running into quota issues today. And you've obviously thought about the trade-offs, so I'm happy with whatever you think is best.

But I think we should revisit it later. I also think it's probably worth adding a `catchWriteDidFail` lifecycle method at some point regardless, so strategy plugin authors can handle these cases individually without potentially affecting plugins they don't own.

jeffposnick replied:
I filed https://github.com/GoogleChrome/workbox/issues/1530 to capture that. To address the immediate quota issues, I'd like to go with a global approach.

philipwalton commented on packages/workbox-core/_private/quota.mjs:
```diff
@@ -0,0 +1,74 @@
+  callbacks.add(callback);
+
+  if (process.env.NODE_ENV !== 'production') {
+    logger.log('Registered a callback to respond to quota errors.', callback);
```
How you checked to see how these function are logged (IIRC normal functions output their `toString()` result)? Maybe it makes more sense to output `.name` value?

jeffposnick replied:
Yeah, it logs the code for the actual callback function. This will normally be

```js
() => this.deleteCacheAndMetadata()
```

or it might be the code for a custom callback, if you write and register your own.

I experimented with this a little bit and could go with just leaving out the logging of the function entirely, but I figured that more context for the development environment logging is better.

Since a developer could theoretically pass in whatever function they want here (in addition to using our automatically registered functions), I don't think we could count on `.name` being meaningful.

jeffposnick replied:
I'd just like to leave it as-is to make sure we're providing the full context to folks who are in debug mode and reading the Console log messages.

philipwalton replied:
SG

philipwalton commented on test/workbox-cache-expiration/node/test-Plugin.mjs:
```diff
@@ -68,6 +69,30 @@ describe(`[workbox-cache-expiration] Plugin`, function() {
+      });
+      plugin.deleteCacheAndMetadata = sandbox.stub();
+
+      await executeCallbacks();
```
Same comment here as above w.r.t the name.

jeffposnick replied:
Not sure what you mean—the name of what?

philipwalton replied:
This is definitely a nit, but I mention it because it confused me a bit when reading the code for the first time. 

Basically my point is that the name `executeCallbacks()` here has no context, so as a reader you have to go searching for where it's defined/imported to figure out what it's doing.

If it were called something like `executeQuotaErrorCallbacks()` or something like that, then it'd be more clear. Alternatively, if you imported the entire namespace as `quotaErrors`, then you'd have more context because it'd be `quotaErrors.executeCallbacks()`.

Anyway, like I said this is a nit. If you think it's fine, I'm OK with it.

philipwalton replied:
Note, this comment primarily applies to [line 90 in `cacheWrapper`](https://github.com/GoogleChrome/workbox/pull/1505/files/b09566d00f13e1e709423d84af609dabd5d89c3d#diff-0e280af723edc4482219279065921834R90), not necessarily here.

jeffposnick replied:
Renamed to `executeQuotaErrorCallbacks()`.

philipwalton commented on packages/workbox-precaching/controllers/PrecacheController.mjs:
```diff
@@ -219,12 +219,14 @@ class PrecacheController {
-    await Promise.all(requests.map(async (request) => {
+    // Process each request/response one at a time, deleting the temporary entry
+    // when done, to help avoid triggering quota errors.
+    for (const request of requests) {
```
+1 to this change. I think it'll make it a lot easier to respond to errors in a predictable way.

philipwalton commented on packages/workbox-core/_private/quota.mjs:
```diff
@@ -0,0 +1,74 @@
+ * a quota error.
+ *
+ * @param {Function} callback
+ *
```
Can you make the spacing here consistent with below. I'm not sure if we prefer a space or not.

jeffposnick replied:
Done.

jeffposnick commented:
I've addressed some of the comments, and provided more context.

Another thing that I'd like to do is write up a "Guide to Handling Quota Errors" or something along those lines for https://developers.google.com/web/tools/workbox/guides/ following the next release, which will explain a bit more about opting-in to the automatic cleanup and registering your own callback.

philipwalton commented on packages/workbox-core/_private/cacheWrapper.mjs:
```diff
@@ -14,12 +14,14 @@
 import pluginUtils from '../utils/pluginUtils.mjs';
+import {WorkboxError} from './WorkboxError.mjs';
+import {assert} from './assert.mjs';
+import {executeCallbacks} from './quota.mjs';
```
Can we also change this name to `executeQuotaErrorCallbacks`?

philipwalton commented:
Thanks Jeff! I left a few random thoughts for the future and one more comment, but overall everything LGTM.

tgangso commented:
When can we expect this to be released?
Thanks
